### ISSUE TITLE
[macOS] Binary gets installed into libdir, crashes with `UI file not found: /opt/local/lib/org.nickvision.tubeconverter/ui/main_window.ui`

### ISSUE THREAD

**Issue by barracuda156** [NickvisionApps/Parabolic#950]
@nlogozzo Looks like the last fix was incomplete: https://github.com/NickvisionApps/Parabolic/blob/8d0e51ea95b6ed316f86b678bcdae50283da90bd/org.nickvision.tubeconverter.gnome/CMakeLists.txt#L1

The build succeeds on macOS now, but does not produce an executable and install fails due to a lack of install target.

```
make: Entering directory `/opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/build'
make: *** No rule to make target `install/fast'.  Stop.
make: Leaving directory `/opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/build'
Command failed:  cd "/opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/build" && /usr/bin/make -w install/fast DESTDIR=/opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/destroot 
Exit code: 2
Error: Failed to destroot parabolic: command execution failed
Error: See /opt/local/var/macports/logs/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/main.log for details.
Error: Follow https://guide.macports.org/#project.tickets if you believe there
is a bug.
Error: Processing of port parabolic failed
```

**Comment by nlogozzo** [comment]
Can you try the latest commit from the `next` branch? https://github.com/NickvisionApps/Parabolic/tree/next

**Comment by nlogozzo** [comment]
You can also specify `-DUI_PLATFORM=gnome` for GTK or `-DUI_PLATFORM=qt` for Qt. 

By default on macOS, `Qt` will be assumed if `UI_PLATFORM` is not specified.

**Comment by barracuda156** [comment]
@nlogozzo Thanks, it fails on this now:
```
:info:build /opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/Parabolic-39a53c7495b9919c3ddd8ddb92b60d0f210226a4/org.nickvision.tubeconverter.gnome/src/views/mainwindow.cpp:132:56: error: too many arguments to function call, expected 0, have 1
:info:build         WindowGeometry geometry{ m_controller->startup(m_controller->getAppInfo().getId() + ".desktop") };
:info:build                                  ~~~~~~~~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
:info:build /opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/Parabolic-39a53c7495b9919c3ddd8ddb92b60d0f210226a4/libparabolic/include/controllers/mainwindowcontroller.h:139:41: note: 'startup' declared here
:info:build         Nickvision::App::WindowGeometry startup();
:info:build                                         ^
:info:build /opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/Parabolic-39a53c7495b9919c3ddd8ddb92b60d0f210226a4/org.nickvision.tubeconverter.gnome/src/views/mainwindow.cpp:182:39: error: too many arguments to function call, expected single argument 'e', have 3 arguments
:info:build         ShellNotification::send(args, m_controller->getAppInfo().getId(), _("Open"));
:info:build         ~~~~~~~~~~~~~~~~~~~~~~~       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
:info:build /opt/local/include/libnick/notifications/shellnotification.h:58:10: note: 'send' declared here
:info:build     void send(const ShellNotificationSentEventArgs& e);
:info:build          ^
:info:build /opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/Parabolic-39a53c7495b9919c3ddd8ddb92b60d0f210226a4/org.nickvision.tubeconverter.gnome/src/views/mainwindow.cpp:416:9: warning: 'gtk_show_uri' is deprecated: Use 'gtk_uri_launcher_launch' instead [-Wdeprecated-declarations]
:info:build         gtk_show_uri(GTK_WINDOW(m_window), helpUrl.c_str(), GDK_CURRENT_TIME);
:info:build         ^
:info:build /opt/local/include/gtk-4.0/gtk/deprecated/gtkshow.h:44:1: note: 'gtk_show_uri' has been explicitly marked deprecated here
:info:build GDK_DEPRECATED_IN_4_10_FOR(gtk_uri_launcher_launch)
:info:build ^
:info:build /opt/local/include/gtk-4.0/gdk/version/gdk-visibility.h:208:39: note: expanded from macro 'GDK_DEPRECATED_IN_4_10_FOR'
:info:build #define GDK_DEPRECATED_IN_4_10_FOR(f) GDK_DEPRECATED_FOR (f)
:info:build                                       ^
:info:build /opt/local/include/gtk-4.0/gdk/version/gdk-visibility.h:31:31: note: expanded from macro 'GDK_DEPRECATED_FOR'
:info:build #define GDK_DEPRECATED_FOR(f) G_DEPRECATED_FOR(f) _GDK_EXTERN
:info:build                               ^
:info:build /opt/local/include/glib-2.0/glib/gmacros.h:1272:44: note: expanded from macro 'G_DEPRECATED_FOR'
:info:build #define G_DEPRECATED_FOR(f) __attribute__((__deprecated__("Use '" #f "' instead")))
:info:build                                            ^
:info:build 1 warning and 2 errors generated.
```

P. S. BTW, forgot to ask: it is possible to avoid requiring `yelp` (as a non-default option)? Or at least somehow do that locally for MacPorts, if it is not something desirable here.
Reason being that it needs `webkit2-gtk`, which broken beyond reasonable repair on multiple macOS versions.

**Comment by nlogozzo** [comment]
Okay try the latest commit in `next`, should now build.

`yelp` is not required. If Parabolic finds `itstool` and `yelp-build` it will use it, if they are not found, it will not use it.

**Comment by barracuda156** [comment]
```
:info:build /opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/Parabolic-1e714ea25418c0943b4e068d0fa822619a7d0cc6/org.nickvision.tubeconverter.gnome/src/views/mainwindow.cpp:189:38: error: expected ';' after expression
:info:build         ShellNotification::send(args)
:info:build                                      ^
:info:build                                      ;
:info:build /opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/Parabolic-1e714ea25418c0943b4e068d0fa822619a7d0cc6/org.nickvision.tubeconverter.gnome/src/views/mainwindow.cpp:424:9: warning: 'gtk_show_uri' is deprecated: Use 'gtk_uri_launcher_launch' instead [-Wdeprecated-declarations]
:info:build         gtk_show_uri(GTK_WINDOW(m_window), helpUrl.c_str(), GDK_CURRENT_TIME);
:info:build         ^
:info:build /opt/local/include/gtk-4.0/gtk/deprecated/gtkshow.h:44:1: note: 'gtk_show_uri' has been explicitly marked deprecated here
:info:build GDK_DEPRECATED_IN_4_10_FOR(gtk_uri_launcher_launch)
:info:build ^
:info:build /opt/local/include/gtk-4.0/gdk/version/gdk-visibility.h:208:39: note: expanded from macro 'GDK_DEPRECATED_IN_4_10_FOR'
:info:build #define GDK_DEPRECATED_IN_4_10_FOR(f) GDK_DEPRECATED_FOR (f)
:info:build                                       ^
:info:build /opt/local/include/gtk-4.0/gdk/version/gdk-visibility.h:31:31: note: expanded from macro 'GDK_DEPRECATED_FOR'
:info:build #define GDK_DEPRECATED_FOR(f) G_DEPRECATED_FOR(f) _GDK_EXTERN
:info:build                               ^
:info:build /opt/local/include/glib-2.0/glib/gmacros.h:1272:44: note: expanded from macro 'G_DEPRECATED_FOR'
:info:build #define G_DEPRECATED_FOR(f) __attribute__((__deprecated__("Use '" #f "' instead")))
:info:build                                            ^
:info:build 1 warning and 1 error generated.
```

**Comment by nlogozzo** [comment]
```
:info:build /opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/Parabolic-1e714ea25418c0943b4e068d0fa822619a7d0cc6/org.nickvision.tubeconverter.gnome/src/views/mainwindow.cpp:189:38: error: expected ';' after expression
```
Ahhh i forgot the `;` 😂😂

Fixed in the new latest commit in `next`!

Sorry doing this all from Windows 🙈

**Comment by nlogozzo** [comment]
`/opt/local/lib/org.nickvision.tubeconverter/org.nickvision.tubeconverter.gnome` this is the executable

Can you build and run the project without trying to install it?

**Comment by barracuda156** [comment]
Either way fails to launch, not finding a component:
```
svacchanda@43-38 ~ % /opt/local/lib/org.nickvision.tubeconverter/org.nickvision.tubeconverter.gnome
[INFO] Started GTK application.
file: /opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/Parabolic-d5182e96739d5a311de0544ec9bfb60bf65c660d/libparabolic/include/controllers/mainwindowcontroller.h(164:108) ``

libc++abi: terminating due to uncaught exception of type std::invalid_argument: UI file not found: /opt/local/lib/org.nickvision.tubeconverter/ui/main_window.ui
zsh: abort      /opt/local/lib/org.nickvision.tubeconverter/org.nickvision.tubeconverter.gnom
svacchanda@43-38 ~ % /opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/build/org.nickvision.tubeconverter.gnome/org.nickvision.tubeconverter.gnome
[INFO] Started GTK application.
file: /opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/Parabolic-d5182e96739d5a311de0544ec9bfb60bf65c660d/libparabolic/include/controllers/mainwindowcontroller.h(164:108) ``

libc++abi: terminating due to uncaught exception of type std::invalid_argument: UI file not found: /opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/build/org.nickvision.tubeconverter.gnome/ui/main_window.ui
zsh: abort
```

**Comment by nlogozzo** [comment]
`UI file not found:` Ahh, you need https://gitlab.gnome.org/jwestman/blueprint-compiler

**Comment by nlogozzo** [comment]
```
/opt/local/bin/org.nickvision.tubeconverter
```

This is also a shortcut to the binary in the `lib` folder. Because the app needs more so it's all provided in `libdir` and then puts a single shortcut in `bindir`

**Comment by barracuda156** [comment]
> `UI file not found:`
> 
> Ahh, you need https://gitlab.gnome.org/jwestman/blueprint-compiler

Perhaps worth adding a check for it? Otherwise the build still succeeds, but the binary is unusable. Anyone installing it will think it is broken.

**Comment by barracuda156** [comment]
> ```
> /opt/local/bin/org.nickvision.tubeconverter
> ```
> 
> This is also a shortcut to the binary in the `lib` folder. Because the app needs more so it's all provided in `libdir` and then puts a single shortcut in `bindir`

Yeah, that works. I will try to make a port for `blueprint-compiler`, it is not there in MacPorts, it seems.

**Comment by nlogozzo** [comment]
> Perhaps worth adding a check for it? Otherwise the build still succeeds, but the binary is unusable. Anyone installing it will think it is broken.

Yeah, I will add a check for it

**Comment by barracuda156** [comment]
Ok, `blueprint-compiler` installed, running the build now breaks at:
```
error: Namespace Adw does not contain a type called Spinner
at /opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/Parabolic-d5182e96739d5a311de0544ec9bfb60bf65c660d/org.nickvision.tubeconverter.gnome/blueprints/add_download_dialog.blp line 27 column 20:
  27 |        child: Adw.Spinner {
     |                   ^^^^^^^
hint: Did you mean `Banner`?
suggestion: replace Spinner with Banner

make[2]: *** [gnome_commands] Error 1
make[2]: Leaving directory `/opt/local/var/macports/build/_opt_svacchanda_SonomaPorts_net_parabolic/parabolic/work/build'
make[1]: *** [org.nickvision.tubeconverter.gnome/CMakeFiles/gnome_commands.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

Do I miss some other dependency?

**Comment by nlogozzo** [comment]
That's libadwaita. But the `Adw.Spinner` comes with version 1.6 you have to make sure you have that.

**Comment by barracuda156** [comment]
Oops, no, we are at 1.5.5. Updating to 1.6.x required updating `gtk4` first, and that is a) a port maintained by someone else and b) a mess to test, since a lot of ports depend on it.

**Comment by nlogozzo** [comment]
> Oops, no, we are at 1.5.5. Updating to 1.6.x requires updating `gtk4` first, and that is a) a port maintained by someone else and b) a mess to test, since a lot of ports depend on it.

Yeah I think trying to get the Qt version might be easier, no?

Besides the GNOME version is really meant to work and function best on Linux. Qt for the rest.

**Comment by barracuda156** [comment]
Qt6 is off question for most of macOS versions, Qt5 is broken on some (which is why I asked about Qt4 earlier).
So no, GTK is doable, just turned out to require substantially more work. Qt is unfeasible until someone fixes Qt6, and I have no time for that.

**Comment by barracuda156** [comment]
I will try updating `gtk4` and `libadwaita` locally, if that works, at least I can confirm if `Parabolic` works on macOS.

**Comment by nlogozzo** [comment]
Ah okay got it. Hopefully it works!

**Comment by nlogozzo** [comment]
I've added messages to cmake about missing dependencies, that you for that suggestion!

**Comment by barracuda156** [comment]
`libadwaita` 1.6.x requires `gtk` 4.15.2+, `gtk` 4.15.2 requires `cairo` 1.18.0. This is endless :) 
The problem is not so much that it requires time to update all the zoo (and then likely need to fix something for less-that-bleeding-edge macOS, because Gnome upstreams don’t bother testing), but that it will be in the limbo indefinitely, if I make a PR into MacPorts, since touching such fundamental dependencies as `cairo` is problematic (and in this case for a good reason). `cairo` is not updated frequently: https://github.com/macports/macports-ports/commits/master/graphics/cairo/Portfile

Maybe it is possible to have a fallback for `libadwaita` 1.5.x?

**Comment by nlogozzo** [comment]
> UPD. Ok, I have an idea, let me see if this can be done with -devel subports, cairo-devel is at 1.18.2, so I will just need to handle gtk4-devel and then make a libadwaita-devel port.

Sounds good! Yeah sorry GTK is dependency hell haha

**Comment by nlogozzo** [comment]
> Maybe it is possible to have a fallback for libadwaita 1.5.x?

This would be hard since there is no Adw.Spinner is libadwaita 1.5 so we'd need like 2 UI files depending on which version one is using which isn't the best idea IMO.

**Comment by barracuda156** [comment]
Nah, gtk 4.15.2 fails to build on macOS Sonoma (i.e. second latest version):
```
FAILED: gtk/libgtk-4.1.dylib 
/usr/bin/clang  -o gtk/libgtk-4.1.dylib  -L/opt/local/lib -I/opt/local/include -Wl,-dead_strip_dylibs -Wl,-headerpad_max_install_names -shared -install_name @rpath/libgtk-4.1.dylib -compatibility_version 1503 -current_version 1503.0 -Wl,-force_load gtk/libgtk.a -Wl,-force_load gtk/css/libgtk_css.a -Wl,-force_load gdk/libgdk.a -Wl,-force_load gsk/libgsk.a -Wl,-headerpad_max_install_names -Wl,-syslibroot,/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk -arch arm64 -pipe -Os -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk -arch arm64 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk -Wl,-rpath,/opt/local/lib gsk/libgsk_f16c.a /opt/local/lib/libgmodule-2.0.dylib /opt/local/lib/libglib-2.0.dylib /opt/local/lib/libintl.dylib /opt/local/lib/libgobject-2.0.dylib /opt/local/lib/libgio-2.0.dylib /opt/local/lib/libpangocairo-1.0.dylib /opt/local/lib/libpango-1.0.dylib /opt/local/lib/libharfbuzz.dylib /opt/local/lib/libcairo.dylib /opt/local/lib/libharfbuzz-subset.dylib /opt/local/lib/libfribidi.dylib /opt/local/lib/libcairo-gobject.dylib /opt/local/lib/libfontconfig.dylib /opt/local/lib/libfreetype.dylib /opt/local/lib/libgdk_pixbuf-2.0.dylib /opt/local/lib/libepoxy.dylib -lm /opt/local/lib/libgraphene-1.0.dylib /opt/local/lib/libXi.dylib /opt/local/lib/libX11.dylib /opt/local/lib/libpangoft2-1.0.dylib -lintl /opt/local/lib/libpng16.dylib /opt/local/lib/libtiff.dylib /opt/local/lib/libjpeg.dylib /opt/local/lib/libXrender.dylib /opt/local/lib/libXext.dylib /opt/local/lib/libXcursor.dylib /opt/local/lib/libXdamage.dylib /opt/local/lib/libXfixes.dylib /opt/local/lib/libXrandr.dylib /opt/local/lib/libXinerama.dylib /opt/local/lib/libcairo-script-interpreter.dylib -lintl
ld: warning: ignoring duplicate libraries: '-lintl'
Undefined symbols for architecture arm64:
  "_gtk_color_picker_quartz_new", referenced from:
      _gtk_color_picker_new in libgtk.a[306](gtkcolorpicker.c.o)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
I can only imagine how broken it is on some earlier systems.

Returning to the question about a fallback for `libadwaita` 1.5.x, is it doable without a huge effort? No big deal if some bells&whistles are not supported as long as the app is still usable.

**Comment by nlogozzo** [comment]
I could just revert Adw.Spinner back into Gtk.Spinner ...

**Comment by nlogozzo** [comment]
Try the latest commit in `next` with 1.5

**Comment by barracuda156** [comment]
> Try the latest commit in `next` with 1.5

@nlogozzo Awesome, that worked:

![parabolic](https://github.com/user-attachments/assets/3555803d-934b-4495-8d45-8dc0125e7b3d)

I will verify on PowerPC within an hour, but at least on the modern macOS it works now.

**Comment by nlogozzo** [comment]
Nice! 🎉🎉🎉

Can you provide me all the steps you used to install all the libraries and the commands to build ? I'd like to create a CI that will build every commit and release automatically

**Comment by barracuda156** [comment]
@nlogozzo Yeah, let me just ensure it works on 10.6 too (so that I can add a fix straightaway if needed).

In the meantime, could you please make a micro-release to incorporate recent patches?

**Comment by nlogozzo** [comment]
> In the meantime, could you please make a micro-release to incorporate recent patches?

I am at work now until later tonight. Tomorrow morning I will.

**Comment by barracuda156** [comment]
There is a minor issue, though looks like nothing about macOS version, but about our usage of `libsecret`:
```
-- Configuring done (40.7s)
CMake Error at /opt/local/lib/cmake/libnick/libnickTargets.cmake:61 (set_target_properties):
  The link interface of target "libnick::libnick" contains:

    PkgConfig::libsecret

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  /opt/local/lib/cmake/libnick/libnickConfig.cmake:27 (include)
  libparabolic/CMakeLists.txt:39 (find_package)


-- Generating done (0.2s)
CMake Warning:
  Manually-specified variables were not used by the project:

    BOOST_INCLUDE_DIR
    BOOST_LIBRARY_DIR
    BOOST_LIB_DIR
    CMAKE_OBJCXX_COMPILER
    CMAKE_OBJC_COMPILER
    CMAKE_POLICY_DEFAULT_CMP0025
    CMAKE_POLICY_DEFAULT_CMP0060


CMake Generate step failed.  Build files cannot be regenerated correctly.
```

Dependencies are there:
```
36-227% port installed libsecret | grep active
  libsecret @0.21.4_0 (active)
36-227% port installed libnick | grep active
  libnick @2024.11.0_0 (active)
```

Any idea what CMake does not like?

**Comment by barracuda156** [comment]
Pkg-config file of `libsecret`:
```
prefix=/opt/local
includedir=${prefix}/include
libdir=${prefix}/lib

exec_prefix=${prefix}
datarootdir=${prefix}/share
datadir=${datarootdir}
sysconfdir=${prefix}/etc

Name: libsecret-1
Description: GObject bindings for Secret Service API
Version: 0.21.4
Requires: glib-2.0 >= 2.44, gio-2.0 >= 2.44, gio-unix-2.0 >= 2.44
Requires.private: libgcrypt >= 1.2.2
Libs: -L${libdir} -lsecret-1
Cflags: -I${includedir}/libsecret-1
```

**Comment by barracuda156** [comment]
If add this to CMakeLists:
```
    pkg_check_modules(secret REQUIRED IMPORTED_TARGET libsecret-1)
    target_link_libraries(${OUTPUT_NAME} PRIVATE libparabolic PkgConfig::gtk PkgConfig::adwaita PkgConfig::libxmlpp PkgConfig::libsecret)
```
it finds `libsecret` normally, but still fails then on the same spot:
```
-- Checking for module 'libsecret-1'
--   Found libsecret-1, version 0.21.4
-- Configuring done (4.7s)
CMake Error at /opt/local/lib/cmake/libnick/libnickTargets.cmake:61 (set_target_properties):
  The link interface of target "libnick::libnick" contains:

    PkgConfig::libsecret

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  /opt/local/lib/cmake/libnick/libnickConfig.cmake:27 (include)
  libparabolic/CMakeLists.txt:39 (find_package)


CMake Error at org.nickvision.tubeconverter.gnome/CMakeLists.txt:23 (target_link_libraries):
  Target "org.nickvision.tubeconverter.gnome" links to:

    PkgConfig::libsecret

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



-- Generating done (0.2s)
```

**Comment by barracuda156** [comment]
Sorry for spamming, fixed now. This was missing, but in another CMakeLists:
```
pkg_check_modules(libsecret REQUIRED IMPORTED_TARGET libsecret-1)
```

**Comment by barracuda156** [comment]
Looks like this fails:
```
    Logger::Logger(const std::filesystem::path& path, LogLevel minimum, bool append)
        : m_minimum{ minimum },
        m_path{ path },
        m_stream{ std::make_shared<std::ofstream>(m_path, append ? std::ios::out | std::ios::app : std::ios::out | std::ios::trunc) }
    {
        if(!m_stream || !m_stream->good())
        {
            throw std::runtime_error("Unable to open log file.");
        }
    }
```

**Comment by barracuda156** [comment]
@nlogozzo Ok, I have no idea how to fix the error correctly and what causes it, however commenting out that line with `throw` in fact allows the app to run.

![parabolic](https://github.com/user-attachments/assets/1ce10475-fe0c-412e-b32e-b94659bc0022)

![parabolic_2](https://github.com/user-attachments/assets/8a28ef10-cb59-4f7b-a6f3-77289f66a565)

I copy-pasted Youtube URL from `smtube`, and downloading works fine.

It would be nice to fix logging properly though :)

**Comment by barracuda156** [comment]
@nlogozzo I have fixed that start-up error now.

**Comment by barracuda156** [comment]
> Can you provide me all the steps you used to install all the libraries and the commands to build ? I'd like to create a CI that will build every commit and release automatically

Last 4 commits in this branch adds whatever was lacking in MacPorts: https://github.com/barracuda156/powerpc-ports/tree/libnick
With that added, `sudo port -v install parabolic` is supposed to produce the app.

**Comment by barracuda156** [comment]
@nlogozzo I think everything here has been sorted, we can close the issue.

P. S. If you happen to be interested, that obscure crash on launch had nothing to do with the code as such (for some reason that `throw` triggered it, but the cause is elsewhere). On older macOS we use a modern `libstdc++` supplied by gcc, so there are two libs of this name, original being in `/usr/lib`. At runtime it may happen that a symbol is picked from the wrong lib, resulting in multiple `malloc` errors (which may or may not crash the running process). Setting `DYDL_LIBRARY_PATH` to point to the right one usually helps. That problem is well-known. Difficulty here was that it did not reach `malloc` errors point, crashing on that logging call. Once I commented out the line, the binary could run and displayed that malloc pattern on startup. Then I just made a custom wrapper to replace yours, which sets `DYDL_LIBRARY_PATH` before invoking the actual binary. That fixed the original problem, so no need to get rid of that `throw` call.
In the portfile code the post-destroot chunk handles that now, replacing the wrapper when C++ runtime is not `libc++`.
